symsrv-fetch.py

```diff
@@ -24,345 +24,468 @@
-                        filename,
-                        id)
-                else:
-                    log.debug('Successfully downloaded %s/%s', filename, id)
-                    file_index.add(key)
-                    sym_file = os.path.join(symbol_path, rel_path)
-                    try:
-                        os.makedirs(os.path.dirname(sym_file))
-                    except OSError:
-                        pass
-                    # TODO: just add to in-memory zipfile
-                    open(sym_file, 'w').write(sym_output)
-            except Win64ProcessError:
-                log.warn('Skipping Win64 PDB: %s/%s' % (filename, id))
-                not_processed_count += 1
-
-    if verbose:
-        sys.stdout.write('\n')
+                log.debug("%s/%s already in skiplist", pdb, debug_id)
+
+    return modules, stats
+
+
+async def collect_info(client, filename, debug_id, code_file, code_id):
+    pdb_path = os.path.join(filename, debug_id, filename)
+    sym_path = os.path.join(filename, debug_id, filename.replace(".pdb", "") + ".sym")
+
+    has_pdb = await server_has_file(client, MICROSOFT_SYMBOL_SERVER, pdb_path)
+    has_code = is_there = False
+    if has_pdb and not await server_has_file(client, MOZILLA_SYMBOL_SERVER, sym_path):
+        has_code = await server_has_file(
+            client, MICROSOFT_SYMBOL_SERVER, f"{code_file}/{code_id}/{code_file}"
+        )
+    elif has_pdb:
+        # if the file is on moz sym server no need to do anything
+        is_there = True
+        has_pdb = False
+
+    return (filename, debug_id, code_file, code_id, has_pdb, has_code, is_there)
+
+
+async def check_x86_file(path):
+    async with AIOFile(path, "rb") as In:
+        chunk = await In.read(32)
+        if chunk.startswith(b"MODULE windows x86 "):
+            return True
+    return False
+
+
+async def run_command(cmd):
+    proc = await asyncio.create_subprocess_shell(
+        cmd, stdout=asyncio.subprocess.PIPE, stderr=asyncio.subprocess.PIPE
+    )
+    _, err = await proc.communicate()
+    err = err.decode().strip()
+
+    return err
+
+
+async def dump_module(
+    output, symcache, filename, debug_id, code_file, code_id, has_code, dump_syms
+):
+    sym_path = os.path.join(filename, debug_id, filename.replace(".pdb", "") + ".sym")
+    output_path = os.path.join(output, sym_path)
+    sym_srv = SYM_SRV.format(symcache)
+
+    if has_code:
+        cmd = f"{dump_syms} {code_file} --code-id {code_id} --store {output} --symbol-server '{sym_srv}' --verbose error"
+    else:
+        cmd = f"{dump_syms} {filename} --debug-id {debug_id} --store {output} --symbol-server '{sym_srv}' --verbose error"
+
+    err = await run_command(cmd)
+
+    if err:
+        log.error(f"Error with {cmd}")
+        log.error(err)
+        return 1
+
+    if not has_code and not await check_x86_file(output_path):
+        log.debug(f"x86_64 binary {code_file}/{code_id} required")
```

Comment:
Do we only support 64 bit?

Reply:
PDB for 32 bits contains everything we need (symbols + stack unwind info).
But PDB for 64 bits don't contain stack unwind info (they're in the binary (.dll/.exe) itself).
So here we're logging because we've a PDB (64 bits) without its DLL/EXE.
(I'll add these lines in the code as a comment)

Reply:
Yeah, a comment would be good here!